packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -490,10 +491,27 @@ export default createRule<Options, MessageId>({
         );
 
         if (isVoidReturningFunctionType(checker, tsNode.name, contextualType)) {
-          context.report({
-            node: node.value,
-            messageId: 'voidReturnProperty',
-          });
+          const signature = checker.getSignatureFromDeclaration(tsNode);
+          if (signature) {
+            const returnType = checker.getReturnTypeOfSignature(signature);
+            if (tsutils.isThenableType(checker, tsNode, returnType)) {
+              const functionNode = node.value;
+              if (isFunction(functionNode)) {
+                if (functionNode.returnType) {
+                  context.report({
+                    node: functionNode.returnType,
+                    messageId: 'voidReturnProperty',
+                  });
+                } else {
+                  context.report({
+                    loc: getFunctionHeadLoc(functionNode, context.sourceCode),
+                    node: functionNode,
+                    messageId: 'voidReturnProperty',
+                  });
+                }
+              }
+            }
+          }
```

Comment:
[Logic] This change is a little misleading to read: it looks like `context.report` is only called in some specific cases. Specifically, if all three `if`s are truthy: `signature`, ``tsutils.isThenableType(...)`, and `isFunction(functionNode)`. But what's really supposed to be happening is changing the reported `loc` or `node` based on some logic.

If we're guaranteed that these cases are happening, then there's no need for `if` statements. I'm inferring they're just here so to help TypeScript narrow down some values, right?

Change request: don't add new, redundant runtime checks.

Hint: I deleted all the added variables except `functionNode` and the tests still passed. So I bet you can remove most of the new `if`s. 🙂

Reply:
Thank you for your feedback! You’re correct; the extra if statements were initially added to assist TypeScript with type narrowing.